File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxProduct = function (root) {
    const MOD = Math.pow(10, 9) + 7;

    // Function to calculate total sum of all nodes
    function sum(root) {
        if (root == null) return 0;
        return (root.val + sum(root.left) + sum(root.right)) % MOD;
    }

    // Function to calculate max product
    function solve(root, totalsum, max) {
        if (root == null) return 0;

        const lsum = solve(root.left, totalsum, max) % MOD;
        const rsum = solve(root.right, totalsum, max) % MOD;

        let res = lsum + rsum + root.val
        max.val = Math.max((totalsum - res) * res, max.val)
        return res;


        // return (root.val + lsum + rsum) % MOD;
    }

    let totalsum = sum(root) % MOD;  // Calculate total sum of the tree
    let max = { val: 0 };

    solve(root, totalsum, max);  // Find max product

    return max.val % MOD;

};

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    long long int MOD = 1000000007;

    int sum(TreeNode* root, int totalsum, long long int& maxproduct) {
        if (root == NULL) {
            return 0;
        }
        long long int lsum = sum(root->left, totalsum, maxproduct) ;
        long long int rsum = sum(root->right, totalsum, maxproduct) ;
        long long int product1 = (root->val % MOD + lsum + rsum) ;
        long long int product2 = (totalsum  - product1  ) ;
        long long int totalproduct = (product1 * product2 ) ;
        maxproduct = max(maxproduct, totalproduct);
        return (root->val  + lsum + rsum) ;
    }
    int maxProduct(TreeNode* root) {
        int totalsum=0;
        long long int maxproduct = 0;
         totalsum = sum(root,totalsum,maxproduct);

        
        long long int val = sum(root, totalsum, maxproduct);
        return maxproduct % MOD;
    }
};